framework/src/main/java/org/checkerframework/framework/flow/CFAbstractStore.java

```diff
@@ -323,22 +341,64 @@ public void insertValue(JavaExpression expr, AnnotationMirror a) {
      *
      * <p>Note that this happens per hierarchy, and if the store already contains information about
      * a hierarchy other than {@code newAnno}'s hierarchy, that information is preserved.
+     *
+     * @param expr an expression
+     * @param newAnno the expression's annotation
      */
-    public void insertOrRefine(JavaExpression expr, AnnotationMirror newAnno) {
+    public final void insertOrRefine(JavaExpression expr, AnnotationMirror newAnno) {
+        insertOrRefine(expr, newAnno, false);
+    }
+
+    /**
+     * Like {@link #insertOrRefine(JavaExpression, AnnotationMirror)}, but permits nondeterministic
+     * expressions to be inserted.
+     *
+     * <p>For an explanation of when to permit nondeterministic expressions, see {@link
+     * #insertValuePermitNondeterministic(JavaExpression, CFAbstractValue)}.
+     *
+     * @param expr an expression
+     * @param newAnno the expression's annotation
+     */
+    public final void insertOrRefinePermitNondeterministic(
+            JavaExpression expr, AnnotationMirror newAnno) {
+        insertOrRefine(expr, newAnno, true);
+    }
+
+    /**
+     * Helper function for {@link #insertOrRefine(JavaExpression, AnnotationMirror)} and {@link
+     * #insertOrRefinePermitNondeterministic}.
+     *
+     * @param expr an expression
+     * @param newAnno the expression's annotation
+     * @param permitNondeterministic whether nondeterministic expressions may be inserted into the
+     *     store
+     */
+    public void insertOrRefine(
```

Comment:
Should this be protected?

Reply:
`insertOrRefine` already existed and was public, as are the other `insert*` methods.  We could change them, but doing so probably belongs in a different pull request as it is orthogonal to this one.

Reply:
There are 3 insertOrRefine methods:
```
public final void insertOrRefine(JavaExpression expr, AnnotationMirror newAnno) {...}
public final void insertOrRefinePermitNondeterministic(
       JavaExpression expr, AnnotationMirror newAnno) {...}
public void insertOrRefine(
       JavaExpression expr, AnnotationMirror newAnno, boolean permitNondeterministic) {...}
```
Only the first method already exists. The last two were added by this pull request. I'm suggesting that the last `insertOrRefine` method should be protected because it should only be called by CFAbstractStore or its subclasses.